### PR TITLE
JSON PDS4 Label

### PR DESCRIPTION
**Pull request by tdddblog**
See https://github.com/NASA-PDS/pds-api/issues/101

**Build Instructions**
- Build `pds-api-javalib` from "swaggerhub-pds_federated_api-pds-api-101" branch.
- Build and install `pds-api-javalib` locally (`mvn install`. You may need to remove "maven-gpg-plugin" from pom.xml if the build fails.)
- Build `registry-api-service`

**Create test data**
- Use the latest Harvest and Registry Manager builds.
- Delete old and create new registry.
- Harvest and load some data into Elasticsearch.
- Latest Harvest version creates `ops:Label_File_Info/ops:json_blob` field (aka JSON BLOB) by default. This field contains PDS4 label in JSON format (compressed and Base64 encoded).
- Make sure that documents in Elasticsearch have JSON BLOB:
  `http://localhost:9200/registry/_search?q=*&_source=ops:Label_File_Info/ops:json_blob`

**Test API Server**
- Start API server (e.g., `java -jar registry-api-service-0.4.0-SNAPSHOT.jar`)
- Go to Swagger UI (`localhost:8080`)
- Select the endpoint you want to test (bundle, collection, product)
- Select the `application/pds4+json` response content type from the dropdown
- Click "Try it out" button
- You should see response similar to this:

```
{
  "summary": {
    "start": 0,
    "limit": 100
  },
  "data": [
    {
      "id": "urn:nasa:pds:kaguya_grs_spectra::1.1",
      "meta": {
        "node_name": "PDS_ENG",
        "label_file": {
          "file_name": "bundle_kaguya_derived.xml",
          "file_ref": "/C:/ws3/Geo/lunar/urn-nasa-pds-kaguya_grs_spectra/bundle_kaguya_derived.xml",
          "creation_date": "2020-11-06T21:09:09Z",
          "file_size": "8556",
          "md5_checksum": "eabfe35431bcfbfc0faa2ced241398dc"
        }
      },
      "pds4": {
        "Product_Bundle": {
          "Identification_Area": {
            "product_class": "Product_Bundle",
            "Modification_History": {
              "Modification_Detail": {
                "modification_date": "2019-09-30",
                "description": "Peer reviewed version of Kaguya Gamma-Ray Spectrometer Corrected Spectra PDS4",
                "version_id": 1
              }
            },
...
```

### CODE REVIEW

**Review by tloubrieu-jpl — CHANGES_REQUESTED** [review]
Hi @tdddblog 
I did not fully test or reviez yet, but now I think that:
- you should use the same business object to get the products fro, elasticSearch
- you should have different methods getProducts for: fullProduct (the pds4+json case) or shortProduct (the existing cases): getFullProduct, getShortProduct because I don't think it is a good idea to have the format specific cases deep into the business object model.
- we should have a mapping somewhere between the mime type relevant for fullproducts, shortProduct (then the kvp format could re-use the shortProduct, at least as a first implementation). The mapping will be used to chose the proper business object request.

That will be useful if we want to re-introduce xml encoding or other alternate format. And mostly I don't want developers to be confused about the usage of the MVC model in the code.

Does that make sense to you ? @jordanpadams

File: src/main/java/gov/nasa/pds/api/engineering/controllers/MyProductsApiBareController.java [hunk 1]
```diff
@@ -178,70 +177,101 @@ protected Products getProducts(String q, String keyword, int start, int limit, L
...
+        {
+            return new ResponseEntity<Object>(HttpStatus.NOT_IMPLEMENTED);
+        }
+    }    
     
-    protected ResponseEntity<Product> getProductResponseEntity(String lidvid){
+    protected ResponseEntity<Object> getProductResponseEntity(String lidvid){
     	String accept = request.getHeader("Accept");
         if ((accept != null) 
         		&& (accept.contains("application/json")
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/api/engineering/controllers/MyProductsApiBareController.java:234]: ok

File: src/main/java/gov/nasa/pds/api/engineering/controllers/MyProductsApiBareController.java [hunk 2]
```diff
@@ -178,70 +177,101 @@ protected Products getProducts(String q, String keyword, int start, int limit, L
...
+                        || accept.contains("application/pds4+json")
+                        || accept.contains("text/html")
+                        || accept.contains("application/xml")
+                        || accept.contains("*/*"))) 
+                        || (accept == null))
+        {
+            try
+            {
+                Object products = null;
+                if("application/pds4+json".equals(accept))
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/api/engineering/controllers/MyProductsApiBareController.java:196]: @tloubrieu-jpl How would you name 3 different getFullProduct() methods which return "pds4+json", "pds4+xml", and key-value pair json? What's wrong with having multiple specialized objects for "pds4+json", "pds4+xml", and standard key-value pair json? It is very hard to maintain one huge class.

**Comment by tloubrieu-jpl** [comment]
@tdddblog different classes can be used for the business objects or services if that makes sense. However we would need a consistent naming for these classes and have them in the single packages. I chose business Object to follow the java design pattern but this might be outdated. If you think Service is better then we should also change the existing business object with this naming.

Otherwise, I really want us to respect the MVC model (see https://docs.spring.io/spring-framework/docs/3.2.x/spring-framework-reference/html/mvc.html#mvc-servlet). The business Object or Service (controller part of the mvc) should return java objects complying with the model generated in pds-api-javalib but without namely referring to the format (xml, json...) which is managed in the view part of the MVC.  That is how springboot is meant to work and how the implementation was initiated to I prefer not to break that.

File: src/main/java/gov/nasa/pds/api/engineering/controllers/MyProductsApiBareController.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/api/engineering/controllers/MyProductsApiBareController.java:196]: @tloubrieu-jpl can you respond to this ASAP? is this something we can merge and then refactor later? or is it something we want to make sure happens right now?

**Comment by tdddblog** [comment]
@tloubrieu-jpl , @jordanpadams, I remove the service class and moved its methods to existing Business object.

File: src/main/java/gov/nasa/pds/api/engineering/controllers/MyProductsApiBareController.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/api/engineering/controllers/MyProductsApiBareController.java:192]: @tloubrieu-jpl, @jordanpadams We should discuss this more. 

(1) I introduced new model classes `Pds4Product`, `Pds4Products`, `Pds4Metadata`, for pds4 mime types, but we should probably try to reuse existing classes `product`, `products`, and `metadata`. Just add new attributes, like `pds4`.

(2) In any case, service layer needs to know which content type was requested, to only get a subset of fields from Elasticsearch (e.g., only "json_blob" for "pds4+json" mime type).

File: src/main/java/gov/nasa/pds/api/engineering/controllers/MyProductsApiBareController.java [hunk 3]
```diff
@@ -178,70 +177,101 @@ protected Products getProducts(String q, String keyword, int start, int limit, L
...
+				|| accept.contains("application/pds4+json"))) {
         	
             try {
             	
             	if (!lidvid.contains("::")) lidvid = this.productBO.getLatestLidVidFromLid(lidvid);
             	
-               	ProductWithXmlLabel product = this.productBO.getProductWithXml(lidvid, this.getBaseURL());
+            	Object product = null;
+            	
+            	if("application/pds4+json".equals(accept))
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/api/engineering/controllers/MyProductsApiBareController.java:243]: @tloubrieu-jpl, @jordanpadams, service layer needs to know which content type was requested, to only get a subset of fields from Elasticsearch (e.g., only "json_blob" for "pds4+json" mime type). So we need either different methods or a content type parameter in getProducts().

Also getProducts() implementation should be in a service layer and not in the controller.

**Comment by jordanpadams** [comment]
@tdddblog I worked through all the merge conflicts (I think) can you look through them and make sure I didn't break anything?

one thing I also bumped into was when I merged the API updates, the changes you made moving the `application/xml` with `object` type did not make it because we removed the `application/xml` response type. I instead replaced that with:
```
            application/pds4+xml:
              schema:
                type: object
```

i tried to fix that here: https://github.com/NASA-PDS/pds-api/commit/3ffc91f8041b23ce0fb16894c9500165d92492c6 . Let me know if that makes sense?

**Comment by jordanpadams** [comment]
@tdddblog I take that back, I am trying to fix these merge conflicts, but I screwed something up either in the API spec or in the merge because I need to change too many things. I hate to do this but you mind taking a look at these merge conflicts? The things we have merged that are bumping into to this:
* NASA-PDS/registry-api-service@3e205b8398f9b89ee36f19be7b1049e7c08f1c79
* NASA-PDS/registry-api-service@b92a03d094cc9403b416b7050953bde177ea3a57
* NASA-PDS/registry-api-service@f787fd1d83f14a3c9ccd6963162bf78d36f59c85

**Comment by tdddblog** [comment]
@jordanpadams I think I merged all conflicts. Also merged yaml in `pds-api-javalib`

File: src/main/java/gov/nasa/pds/api/engineering/elasticsearch/ElasticSearchConfig.java [hunk 4]
```diff
@@ -90,11 +89,17 @@ public ElasticSearchRegistryConnection ElasticSearchRegistryConnection() {
 	
 	@Bean("productBO")
 	public ProductBusinessObject ProductBusinessObject() {
-		
-		
 		return new ProductBusinessObject(this.ElasticSearchRegistryConnection());
 	}
+
 	
+    @Bean("pds4JsonProductService")
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/api/engineering/elasticsearch/ElasticSearchConfig.java]: I removed it

**Comment by jordanpadams** [comment]
@tdddblog unfortunately, looks like I am unable to compile the current branch:

```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /Users/jpadams/Documents/proj/pds/pdsen/workspace/registry-api-service/src/main/java/gov/nasa/pds/api/engineering/controllers/MyCollectionsApiController.java:[38,8] gov.nasa.pds.api.engineering.controllers.MyCollectionsApiController is not abstract and does not override abstract method bundlesContainingCollection(java.lang.String,java.lang.Integer,java.lang.Integer,java.util.List<java.lang.String>,java.util.List<java.lang.String>,java.lang.Boolean) in gov.nasa.pds.api.base.CollectionsApi
[ERROR] /Users/jpadams/Documents/proj/pds/pdsen/workspace/registry-api-service/src/main/java/gov/nasa/pds/api/engineering/controllers/MyCollectionsApiController.java:[201,37] bundlesContainingCollection(java.lang.String,@javax.validation.Valid java.lang.Integer,@javax.validation.Valid java.lang.Integer,@javax.validation.Valid java.util.List<java.lang.String>,@javax.validation.Valid java.util.List<java.lang.String>,@javax.validation.Valid java.lang.Boolean) in gov.nasa.pds.api.engineering.controllers.MyCollectionsApiController cannot implement bundlesContainingCollection(java.lang.String,java.lang.Integer,java.lang.Integer,java.util.List<java.lang.String>,java.util.List<java.lang.String>,java.lang.Boolean) in gov.nasa.pds.api.base.CollectionsApi
  return type org.springframework.http.ResponseEntity<gov.nasa.pds.model.Products> is not compatible with org.springframework.http.ResponseEntity<java.lang.Object>
[ERROR] /Users/jpadams/Documents/proj/pds/pdsen/workspace/registry-api-service/src/main/java/gov/nasa/pds/api/engineering/controllers/MyCollectionsApiController.java:[200,5] method does not override or implement a method from a supertype
```

I tried this both with the `pds-api-javalib` `main` and `swaggerhub-pds_federated_api-pds-api-101` branches (got way more errors on the `swaggerhub-pds_federated_api-pds-api-101` branch)

**Review by jordanpadams — APPROVED** [review]
going to merge this for now to try to cleanup all these merge conflicts, but i think something is still broken.

**Comment by tdddblog** [comment]
@jordanpadams 
I was able to reproduce it. That's the way mvn handles snapshots. To compile you have to:
(1) `mvn build` registry-api-service. It will download latest snapshot from `main` branch and fail.
(2) `mvn install` pds-api-javalib. It will create local snapshot of ...pds-api-101 branch.
(3) Run `mvn build` in registry-api-service second time. This time it should build

**Pull request merged**